src/box/memtx_engine.cc

```diff
@@ -631,7 +633,9 @@ memtx_engine_commit(struct engine *engine, struct txn *txn)
 	struct txn_stmt *stmt;
 	stailq_foreach_entry(stmt, &txn->stmts, next) {
 		if (memtx_tx_manager_use_mvcc_engine) {
-			assert(stmt->space->engine == engine);
+			assert(stmt->space->engine == stmt->engine);
+			if (stmt->engine != engine)
+				continue;
```

Comment:
There shouldn't be any alien tuples in this case AFAIU. At the same time, I don't think we should call `memtx_space_upgrade_untrack_tuple` below for alien tuples.

Reply:
Changed to:
```diff
@@ -631,10 +632,11 @@ memtx_engine_commit(struct engine *engine, struct txn *txn)
        struct txn_stmt *stmt;
        stailq_foreach_entry(stmt, &txn->stmts, next) {
                if (memtx_tx_manager_use_mvcc_engine) {
+                       assert(stmt->engine == engine);
                        assert(stmt->space->engine == engine);
                        memtx_tx_history_commit_stmt(stmt);
                }
-               if (stmt->engine_savepoint != NULL) {
+               if (stmt->engine == engine && stmt->engine_savepoint != NULL) {
                        struct space *space = stmt->space;
                        struct tuple *old_tuple = stmt->rollback_info.old_tuple;
                        if (space->upgrade != NULL && old_tuple != NULL)
```